web_client/views/imageViewerWidget/geojs.js

```diff
@@ -52,47 +70,129 @@ var GeojsImageViewerWidget = ImageViewerWidget.extend({
+     * @returns {Promise}
+     */
     drawRegion: function (model) {
-        if (!this.viewer) {
-            return;
-        }
-        var layer = this.viewer.createLayer('annotation');
-        layer.geoOn(
-            window.geo.event.annotation.state,
-            (evt) => {
-                var annotation = evt.annotation;
-                var left, top, width, height, c;
-                if (annotation.type() === 'rectangle') {
-                    c = annotation.coordinates();
-                    left = Math.round(c[0].x);
-                    top = Math.round(c[1].y);
-                    width = Math.round(c[2].x - left);
-                    height = Math.round(c[3].y - top);
-                    model.set('value', [left, top, width, height], {trigger: true});
-                    window.setTimeout(() => this.viewer.deleteLayer(layer), 10);
+        model = model || new Backbone.Model();
+        return this.startDrawMode('rectangle', {trigger: false}).then((elements) => {
+            /*
+             * Strictly speaking, the rectangle drawn here could be rotated, but
+             * for simplicity we will set the region model assuming it is not.
+             * To be more precise, we could expand the region to contain the
+             * whole rotated rectangle.  A better solution would be to add
+             * a draw parameter to geojs that draws a rectangle aligned with
+             * the image coordinates.
+             */
+            var element = elements[0];
+            var width = Math.round(element.width);
+            var height = Math.round(element.height);
+            var left = Math.round(element.center[0] - element.width / 2);
+            var top = Math.round(element.center[1] - element.height / 2);
+
+            model.set('value', [
+                left, top, width, height
+            ], {trigger: true});
+
+            return model.get('value');
+        });
+    },
+
+    /**
+     * Set the image interaction mode to draw the given type of annotation.
+     *
+     * @param {string} type An annotation type
+     * @param {object} [options]
+     * @param {boolean} [options.trigger=true]
+     *      Trigger a global event after creating each annotation element.
+     * @returns {Promise}
+     *      Resolves to an array of generated annotation elements.
+     */
+    startDrawMode: function (type, options) {
+        var layer = this.annotationLayer;
+        var elements = [];
+        var annotations = [];
+
+        layer.mode(null);
+        layer.geoOff(window.geo.event.annotation.state);
+        layer.removeAllAnnotations();
+        return new Promise((resolve) => {
+            var element;
+
+            options = _.defaults(options || {}, {trigger: true});
+            layer.geoOn(
+                window.geo.event.annotation.state,
+                (evt) => {
+                    element = convertAnnotation(evt.annotation);
+                    if (!element.id) {
```

Comment:
The reason this is done here is because annotation element models are generated for the draw widget before they are submitted to the server.  This "id" is really only used to uniquely identify the elements in the [DrawWidget](https://github.com/DigitalSlideArchive/HistomicsTK/pull/296/files#diff-94336ac7f0b9fbbea193e4bf777a6c91R98) as a "view model".  Because there aren't any endpoints for the annotation element model, it isn't useful for communication with the server (right now anyway).

I took the approach that drawn elements are temporary until submitted as a group into an annotation.  If we had endpoints such as `POST /annotation/:id/element` and `PUT /annotation/:id/element/:id` to set the individual elements, it would make more sense to make the `ElementModel` a server-backed backbone model and POST after creation to set the id.  I could do so now, but it would take some reorganization of how the annotation models are created in the UI.